### ISSUE TITLE
Can we inject modules to .ipynb?

### ISSUE THREAD

**Issue by AkhilSinghRana** [nteract/papermill#377]
This is more of a question. Is it possible to inject objects or modules to an existing notebook? I wanted to pass an object to a cell. This object then can be used to call different functions.

**Comment by MSeal** [comment]
Sort of.

Papermill support basic json data types being transferable because it's easy to translate them to the underlying notebook execution process by translating the data to code in the notebook. Since the notebook doesn't execute in the same process as papermill it's challenging to do direct object transfers, especially since the underlying kernel doesn't even necessarily run the same language (or language version).

What you can do is [extend translators](https://papermill.readthedocs.io/en/latest/extending-overview.html) to convert some object types to more advanced data types in the parameter cell of the notebook. Some systems do this to enable loading internal data structures at runtime when a data path is provided or some other complex object. What isn't possible is to pass arbitrary in-memory objects to the notebook execution unless they're json friendly.

**Comment by AkhilSinghRana** [comment]
Yes, makes sense. I was getting error that said modules can't be pickled. Can we not serialize the objects to disk and then deserialize it from notebook?

**Comment by MSeal** [comment]
Possibly under specialized circumstances, but remember that the kernel is not guaranteed to be the same python version with the same dependencies available. Most pickled objects don't work across python versions, and the kernel might not even be in python at all. So as a general implementation, no. For specific use cases where you know they're always going to be the same, then maybe. It's probably still an antipattern and prone to failure whenever anything changes.
